Orange/widgets/settings.py

```diff
@@ -169,16 +164,12 @@ def encode_setting(cls, context, setting, value):
             else:
                 return copy.copy(value)
 
-        if isinstance(setting, ContextSetting):
-            if isinstance(value, str):
-                variables = {
-                    **({} if setting.exclude_attributes else context.attributes),
-                    **({} if setting.exclude_metas else context.metas)}
-                if value in variables:
-                    cls._warn_about_str_var_settings(setting)
-                    return value, variables[value]
-            elif isinstance(value, Variable):
+        if isinstance(value, Variable):
+            if isinstance(setting, ContextSetting):
                 return cls.encode_variable(value)
+            else:
+                raise ValueError("Variables must be stored as ContextSettings; "
+                                 f"change {setting.name} to ContextSetting.")
```

Comment:
Seems ok to me, but I have no idea about all the types of `value` we support and need to handle here.
And because this looks like it is not just removing a deprecation, but changing functionality - somebody please confirm this is intended and OK.

1) If it was a ContextSetting and str it gave a deprecation before, now it goes to the below default `return copy...`
What happens in that case? After removing deprecation warnings I would expect old code to get some error that still makes it clear that things changed and this is not supported anymore.

2) If it was a Variable and not a ContextSetting before it went to the below default `return copy...`. After this it will raise the new ValueError. Is this a fix for something bundled with the removal of the deprecation?

Reply:
1. Yes, there are two options: when we removing support for something we can either raise an exception or change the behaviour. In this case, it changes the behaviour that, in my opinion, should be changed because it was ugly and dangerous. Plus, raising an exception doesn't work correctly.

    Before, it worked like this. You had a context setting and it stored a string `"foo"`. The handler didn't know if *foo* is a name of a variable or just a string, so it wouldn't know how to save it. The only thing it could do was to check whether there existed a variable named *foo* and if it did, it assumed the setting "foo" referred to that variable. So it encoded the string as a (variable name, variable type) pair. (And, for the last few years, also gave a warning.) This was obviously not foolproof -- there could exist a variable whose name would coincidentally match an unrelated string setting. But even if this happened, the widget wouldn't notice since the variable was decoded back to string `"foo"`. (The only side effect would be that a context without variable `"foo"` wouldn't match.)

    With the current version, that is, coding as a string, domain context matching (in incorrectly implemented widgets) would cease to observe this variable, possibly resulting in an error due to mismatched context. While errors should generally be detected sooner (=here) than later (=some time after `openContext`), I'd in this case prefer later.

    Why? If we change the code so that it would raise an exception, we would get an exception whenever a name of some variable coincidentally matched a string setting (and these kinds of errors did not happen before this PR!). That is, we would in essence have "invalid variable names" -- if there is a string setting `"foo"`, the widget would fail it the domain contained a variable `"foo"`.

    In short, the only benefit of raising an exception here is to raise it at the point of the problem and not later in the code. The downside is that detection of a problem has false positive and it puts arbitrary restraints on variable names, that result in raising strange exception to the end-users.


2. No, this is unrelated, it's just something I discovered when simplifying the conditions. If a `Variable` was stored in a non-context setting, it would be encoded as a value and pickled. After loading a schema, we would probably get an error because the widget's setting would contain a reference to a variable that does not appear in the domain. So it is better to raise an exception when saving the schema. This is thus not really a new exception, just an exception that can be (reliable) shown earlier.